File: C#/Day5.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Collections.Generic;
using System.Linq;
using System.Collections.Generic;
using System.Linq;

namespace AdventSolutions
{
    /// <summary>
    /// Day 5 solution: Checking ordering rules on updates and topologically sorting them.
    /// </summary>
    public static class Day5
    {
        private static (List<(int, int)> Rules, List<List<int>> Updates) ParseInput(string filename)
        {
            var lines = File.ReadAllLines(filename).Select(l => l.Trim()).ToList();
            int blankIndex = lines.IndexOf("");

            var ruleLines = lines.Take(blankIndex).Where(x => !string.IsNullOrWhiteSpace(x)).ToList();
            var updateLines = lines.Skip(blankIndex + 1).Where(x => !string.IsNullOrWhiteSpace(x)).ToList();

            var rules = new List<(int, int)>();
            foreach (var line in ruleLines)
            {
                var parts = line.Split('|', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length == 2 && int.TryParse(parts[0].Trim(), out int x) && int.TryParse(parts[1].Trim(), out int y))
                {
                    rules.Add((x, y));
                }
            }

            var updates = new List<List<int>>();
            foreach (var line in updateLines)
            {
                var pages = line.Split(',', StringSplitOptions.RemoveEmptyEntries)
                                .Select(s => int.Parse(s.Trim())).ToList();
                updates.Add(pages);
            }

            return (rules, updates);
        }

        private static bool IsCorrectlyOrdered(List<int> update, List<(int,int)> rules)
        {
            var pagePositions = update.Select((p, i) => (p, i)).ToDictionary(x => x.p, x => x.i);
            foreach (var (x, y) in rules)
            {
                if (pagePositions.ContainsKey(x) && pagePositions.ContainsKey(y))
                {
                    if (pagePositions[x] > pagePositions[y])
                        return false;
                }
            }
            return true;
        }

        private static List<int> TopologicalSort(HashSet<int> pages, List<(int,int)> relevantRules)
        {
            var graph = new Dictionary<int, List<int>>();
            var inDegree = new Dictionary<int, int>();
            foreach (var p in pages)
            {
                graph[p] = new List<int>();
                inDegree[p] = 0;
            }

            foreach (var (x, y) in relevantRules)
            {
                graph[x].Add(y);
                inDegree[y]++;
            }

            var queue = new Queue<int>(inDegree.Where(kv => kv.Value == 0).Select(kv => kv.Key));
            var sorted = new List<int>();
            while (queue.Count > 0)
            {
                int node = queue.Dequeue();
                sorted.Add(node);
                foreach (var neighbor in graph[node])
                {
                    inDegree[neighbor]--;
                    if (inDegree[neighbor] == 0)
                        queue.Enqueue(neighbor);
                }
            }

            return sorted;
        }

        private static int PartOne(List<(int,int)> rules, List<List<int>> updates)
        {
            int total = 0;
            foreach (var upd in updates)
            {
                if (IsCorrectlyOrdered(upd, rules))
                {
                    int middleIndex = upd.Count / 2;
                    total += upd[middleIndex];
                }
            }
            return total;
        }

        private static int PartTwo(List<(int,int)> rules, List<List<int>> updates)
        {
            int total = 0;
            var incorrectUpdates = updates.Where(upd => !IsCorrectlyOrdered(upd, rules)).ToList();

            foreach (var upd in incorrectUpdates)
            {
                var pageSet = upd.ToHashSet();
                var relevantRules = rules.Where(r => pageSet.Contains(r.Item1) && pageSet.Contains(r.Item2)).ToList();
                var sortedUpd = TopologicalSort(pageSet, relevantRules);
                int middleIndex = sortedUpd.Count / 2;
                total += sortedUpd[middleIndex];
            }

            return total;
        }

        public static void Run()
        {
            string filePath = "../data/Day5.txt";
            if (!File.Exists(filePath))
            {
                Console.WriteLine("Day5 data file not found.");
                return;
            }

            var (rules, updates) = ParseInput(filePath);

            // Part One
            int partOneResult = PartOne(rules, updates);
            Console.WriteLine("Day 5 - Part 1 Result: " + partOneResult);

            // Part Two
            int partTwoResult = PartTwo(rules, updates);
            Console.WriteLine("Day 5 - Part 2 Result: " + partTwoResult);
        }
    }
}

File: python/Day5.py
from collections import defaultdict, deque
from typing import List, Tuple

def parse_input(filename: str) -> Tuple[List[Tuple[int,int]], List[List[int]]]:
    """
    Parse the puzzle input file which contains ordering rules followed by updates.

    Parameters:
        filename (str): The path to the input file.

    Returns:
        Tuple[List[Tuple[int,int]], List[List[int]]]: A tuple containing the rules
        (as a list of (x, y) tuples) and updates (as a list of lists of page numbers).
    """
    with open(filename, "r") as f:
        lines = [line.strip() for line in f]

    blank_index = lines.index('')
    rule_lines = lines[:blank_index]
    update_lines = lines[blank_index+1:]

    rules = []
    for line in rule_lines:
        if line:
            x, y = line.split('|')
            x, y = x.strip(), y.strip()
            rules.append((int(x), int(y)))

    updates = []
    for line in update_lines:
        if line:
            pages = [int(x.strip()) for x in line.split(',')]
            updates.append(pages)

    return rules, updates

def is_correctly_ordered(update: List[int], rules: List[Tuple[int,int]]) -> bool:
    """
    Check if an update's pages are in the correct order based on the given rules.

    Parameters:
        update (List[int]): The list of page numbers in the update.
        rules (List[Tuple[int,int]]): The ordering rules as (x, y) pairs.

    Returns:
        bool: True if the update is correctly ordered, False otherwise.
    """
    page_positions = {page: i for i, page in enumerate(update)}
    for (x, y) in rules:
        if x in page_positions and y in page_positions:
            if page_positions[x] > page_positions[y]:
                return False
    return True

def topological_sort(pages: set, relevant_rules: List[Tuple[int,int]]) -> List[int]:
    """
    Perform a topological sort on the given pages based on the relevant_rules.

    Parameters:
        pages (set): The set of pages to order.
        relevant_rules (List[Tuple[int,int]]): Rules that apply to these pages.

    Returns:
        List[int]: A valid ordering of the pages that satisfies all rules.
    """
    graph = defaultdict(list)
    in_degree = {p: 0 for p in pages}

    for x, y in relevant_rules:
        graph[x].append(y)
        in_degree[y] += 1

    queue = deque([node for node in pages if in_degree[node] == 0])
    sorted_order = []
    while queue:
        node = queue.popleft()
        sorted_order.append(node)
        for neighbor in graph[node]:
            in_degree[neighbor] -= 1
            if in_degree[neighbor] == 0:
                queue.append(neighbor)

    return sorted_order

def part_one(rules: List[Tuple[int,int]], updates: List[List[int]]) -> int:
    """
    Sum the middle page of each correctly-ordered update.

    Parameters:
        rules (List[Tuple[int,int]]): The ordering rules.
        updates (List[List[int]]): The list of updates.

    Returns:
        int: The sum of the middle pages of all correctly-ordered updates.
    """
    total = 0
    for upd in updates:
        if is_correctly_ordered(upd, rules):
            middle_index = len(upd) // 2
            total += upd[middle_index]
    return total

def part_two(rules: List[Tuple[int,int]], updates: List[List[int]]) -> int:
    """
    For each incorrectly-ordered update, find the correct order and sum their middle pages.

    Parameters:
        rules (List[Tuple[int,int]]): The ordering rules.
        updates (List[List[int]]): The list of updates.

    Returns:
        int: The sum of the middle pages of all re-ordered incorrect updates.
    """
    incorrect_updates = [upd for upd in updates if not is_correctly_ordered(upd, rules)]

    total = 0
    for upd in incorrect_updates:
        page_set = set(upd)
        relevant_rules = [(x, y) for (x, y) in rules if x in page_set and y in page_set]

        sorted_upd = topological_sort(page_set, relevant_rules)
        middle_index = len(sorted_upd) // 2
        total += sorted_upd[middle_index]

    return total

def main():
    """Execute Day 5 puzzle solution."""
    rules, updates = parse_input("../data/Day5.txt")

    # Part One:
    part_one_result = part_one(rules, updates)
    print("Day 5 - Part 1 Result:", part_one_result)

    # Part Two:
    part_two_result = part_two(rules, updates)
    print("Day 5 - Part 2 Result:", part_two_result)

if __name__ == "__main__":
    main()
